lib/BasicAnalyses/CSVAliasAnalysis.cpp

```diff
@@ -0,0 +1,259 @@
+/// \file CSVAliasAnalysis.cpp
+/// \brief Decorate memory accesses with information about CSV aliasing, and
+///        optionally segregate direct stack accesses from all other memory
+///        accesses through additional alias information.
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/BasicAnalyses/CSVAliasAnalysis.h"
+
+using namespace llvm;
+
+class SegregateDirectStackAccesses {
+  CSVAliasAnalysisInterface *CSVAA;
+  std::vector<Instruction *> DirectStackAccesses;
+  std::vector<Instruction *> NotDirectStackAccesses;
+
+public:
+  SegregateDirectStackAccesses(CSVAliasAnalysisInterface *CSVAA) :
+    CSVAA(CSVAA) {}
+  void segregateAccesses(Function &F);
+
+private:
+  void decorateStackAccesses(Function &F);
+};
+
+// Inttoptr instructions basically inhibits all optimizations. In particular,
+// when an integer is inttoptr'd twice with different destination type, alias
+// analysis messes up. Hence, ensure that an inttoptr already exists before
+// emitting a new one, when canonicalizing adds into geps later.
+static Value *createOpaqueIntToPtr(IRBuilder<> &Builder, Value *V) {
+  Value *ITP = nullptr;
+  for (auto *U : V->users()) {
+    if (auto *I = dyn_cast<IntToPtrInst>(U)) {
+      ITP = I;
+      break;
+    }
+  }
+
+  if (!ITP)
+    return Builder.CreateIntToPtr(V, Builder.getInt8PtrTy());
+  return Builder.CreateBitCast(ITP, Builder.getInt8PtrTy());
+}
+
+static void canonicalizeIntoGEP(Value *SP, Instruction *I) {
+  ConstantInt *Offset = nullptr;
+  Value *Ptr = nullptr;
+
+  if (skipCasts(Ptr = I->getOperand(0)) == SP)
+    Offset = dyn_cast<ConstantInt>(I->getOperand(1));
+  else if (skipCasts(Ptr = I->getOperand(1)) == SP)
+    Offset = dyn_cast<ConstantInt>(I->getOperand(0));
+
+  if (Ptr && Offset) {
+    IRBuilder<> Builder(I);
+    auto *ITP = createOpaqueIntToPtr(Builder, Ptr);
+    auto *GEP = Builder.CreateGEP(Builder.getInt8Ty(), ITP, Offset);
+    auto *PTI = Builder.CreatePtrToInt(GEP, I->getType());
+    I->replaceAllUsesWith(PTI);
+  }
+}
+
+template<bool SegregateStackAccesses>
+using CSVAAP = CSVAliasAnalysisPass<SegregateStackAccesses>;
+
+template<bool SegregateStackAccesses>
+void CSVAAP<SegregateStackAccesses>::initializeAliasInfo(Module &M) {
+  LLVMContext &Context = M.getContext();
+  QuickMetadata QMD(Context);
+  MDBuilder MDB(Context);
+
+  AliasDomain = MDB.createAliasScopeDomain("revngAliasDomain");
```

Comment:
```suggestion
  AliasDomain = MDB.createAliasScopeDomain("revng.alias.domain");
```

Also `revng.alias.domain` is too generic. If it's about CSVs mention it in the alias domain name. Same if it's about stack accesses.

Reply:
I'm changing it to `CSVAliasDomain`. The dots notation is used by metadatas, this is a scope domain.